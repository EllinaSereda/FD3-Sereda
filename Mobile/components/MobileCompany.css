/*.MobileCompany table
{
margin:0 auto;
width:100%;
border-collapse: collapse;
}
.MobileCompany td,th
{
  width: 14.2%;
  padding:10px;
  border:1px solid black;
}
.MobileCompanyName
{
  font-weight: bold;
}

.MobileCompanyClients
{
  padding-top: 5px;
}*/
