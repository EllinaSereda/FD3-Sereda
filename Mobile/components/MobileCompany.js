import React from 'react';
import PropTypes from 'prop-types';

import MobileClient from './MobileClient';
import Card from './Card';
import {voteEvents} from './events';

import './MobileCompany.css';

class MobileCompany extends React.PureComponent {

  static propTypes = {
    name: PropTypes.string.isRequired,
    clients:PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        fam: PropTypes.string.isRequired,
        im: PropTypes.string.isRequired,
        otch: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
      })
    ),
  };
  
  state = {
    name: this.props.name,
    clients: this.props.clients,
    status:0, //0 - все, 1 - активные , 2 - заблокированные
    CardMode:0,
    selectCli:null,
    newItemID:this.props.clients.length+1,
  };
  componentDidMount = () => {
    voteEvents.addListener('EDeleted',this.deleteClient);  //какого клиента удалили
    voteEvents.addListener('EChanged',this.changeClient);   //измененные данные клиента
    voteEvents.addListener('ECard',this.changeCardMode);   //режим карты 
    voteEvents.addListener('EEditCli',this.selectCli);  //клиент на редактирование
    voteEvents.addListener('ENew',this.addToCliends); //
  };

componentWillUnmount = () => {
    voteEvents.removeListener('EDeleted',this.deleteClient);
    voteEvents.removeListener('EChanged',this.changeClient);
    voteEvents.removeListener('ECard',this.changeCardMode);
    voteEvents.removeListener('EEditCli',this.selectCli);
    voteEvents.removeListener('ENew',this.addToCliends);
  };

changeCardMode=(code)=>{
    this.setState({CardMode:code});
  }
selectCli = (id) =>{  //Какой клиент на редактировании
  this.state.clients.forEach(v=>{(id==v.id)?this.setState({selectCli:v}):null});
}

deleteClient = (code) =>{
    this.setState({clients:this.state.clients.filter(v=>v.id!=code)},this.setNewId)
  
}
setNewId = () =>{
  this.setState({newItemID:this.state.clients.length+1});
}
addToCliends = (client) =>{
  let newClients=[...this.state.clients]; // копия самого массива клиентов
  newClients.push(client);
  this.setState({clients:newClients});
  this.setState({newItemID:(this.state.newItemID+1)});
  this.changeCardMode(0);

}

addClient=()=>{  //добавление клиента 
  this.changeCardMode(2);
}


changeClient = (client) => {   //Редактирование клиента
  let changed=false;
  let newClients=[...this.state.clients]; // копия самого массива клиентов
  newClients.forEach( (c,i) => {
    if ( c.id==client.id ) {
      let newClient={...c}; // копия хэша изменившегося клиента
      newClient=client;
      newClients[i]=newClient;
      changed=true;
    }
  } )
  if ( changed ){
    this.setState({clients:newClients});
  }
  this.changeCardMode(0);
}


  setName1 = () => {
    this.setState({name:'МТС'});
  };

  setName2 = () => {
    this.setState({name:'Velcom'});
  };
  setAll = () =>{  //Показывать всех
    this.setStatus(0);
  }
  setAct = () =>{ //Только активных
    this.setStatus(1);
  }
  setBlock = () =>{  //только заблокированных
    this.setStatus(2);
  }

  setStatus = (mode) =>{
    let changed=false;
    if (this.state.status!=mode){
      changed=true;
    }
    if ( changed ){
      this.setState({status:mode});
    }
  }


  
  render() {

    console.log("MobileCompany render");

    var clientsCode=this.state.clients.map( client => {
      let code=null;
        switch (this.state.status){
          case 0:
          code=<MobileClient  key={client.id} info={client} />;
          break;
          case 1:
          if (client.balance>=0){
            code=<MobileClient key={client.id} info={client} />;
          }
          break;
          case 2:
          if (client.balance<0){
            code=<MobileClient key={client.id} info={client} />;
          }
          break;
        }
        return code;
      }
    );

    return (
      <div className='MobileCompany'>
        <input type="button" value="Velcom" onClick={this.setName2} />
        <input type="button" value="МТС" onClick={this.setName1} />
        <div className='MobileCompanyName'>Компания &laquo;{this.state.name}&raquo;</div>
        <input type="button" value="Все" onClick={this.setAll} />
        <input type="button" value="Активные" onClick={this.setAct} />
        <input type="button" value="Заблокированные"  onClick={this.setBlock} />
        <table>
          <thead>
            <tr>
              <th>Фамилия</th>
              <th>Имя</th>
              <th>Отчество</th>
              <th>Баланс</th>
              <th>Статус</th>
              <th>Редактировать</th>
              <th>Удалить</th>
            </tr>
          </thead>
          <tbody>{clientsCode}</tbody>
        </table>
      
        <input type="button" value="Добавить клиента" onClick={this.addClient} />
        <Card
          CardMode={this.state.CardMode}
         newItemID={this.state.newItemID}
         info={this.state.selectCli}
        />
      </div>
    )
    ;

  }

}

export default MobileCompany;
