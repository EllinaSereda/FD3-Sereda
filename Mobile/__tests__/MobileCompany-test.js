"use strict";

import React from 'react';
import renderer from 'react-test-renderer';

import MobileCompany from '../components/MobileCompany';
//import MobileClient from '../components/MobileClient';

test('работа MobileCompany', () => {
    let companyName='Velcom';
    let clientsArr=[ 
      {id:1, fam:"Иванов", im:"Иван", otch:"Иванович", balance:200}, 
      {id:2, fam:"Сидоров", im:"Сидор", otch:"Сидорович", balance:250}, 
      {id:3, fam:"Петров", im:"Пётр", otch:"Петрович", balance:180},
      {id:4, fam:"Григорьев", im:"Григорий", otch:"Григорьевич", balance:220},
    ];
  // создаём тестовую версию компонента
  const component = renderer.create(
    <MobileCompany 
    name={companyName}
    clients={clientsArr}
  />
  );
  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  
  // найдём в вёрстке компонента количество продуктов в таблице
  const Spisok1=(component.root.find( el => el.type=='tbody')).children.length;
  // найдём в вёрстке компонента кнопку удалить для первого продукта 
  const Delete=component.root.findAll( el =>( el.props.value=='Удалить'))[0];
  //нажмем на кнопку
  Delete.props.onClick();
  // найдём в вёрстке компонента количество продуктов в таблице
  const Spisok2=(component.root.find( el => el.type=='tbody')).children.length;
  //проверим уменьшилось ли количество  элементов в таблице
  expect(Spisok2).toBe(Spisok1-1);


  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
  
  // найдём в вёрстке компонента кнопку МТС
  const   MTC=component.root.find(el =>( el.props.value=='МТС'));
  //нажмем на кнопку, изменим название компании
  MTC.props.onClick();

  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
 
  const Name=(component.root.find( el => el.props.className=='MobileCompanyName')).children;
  expect(Name[1]).toBe('МТС');

  const   Blocked=component.root.find(el =>( el.props.value=='Заблокированные'));
  Blocked.props.onClick();

  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  const spisok=(component.root.find( el => el.type=='tbody')).children.length;
  expect(spisok).toBe(0);
    
});
