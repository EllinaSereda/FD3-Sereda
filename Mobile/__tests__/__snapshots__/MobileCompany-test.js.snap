// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа MobileCompany 1`] = `
<div
  className="MobileCompany"
>
  <input
    onClick={[Function]}
    type="button"
    value="Velcom"
  />
  <input
    onClick={[Function]}
    type="button"
    value="МТС"
  />
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <input
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="MobileClient"
      >
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="active"
        >
          Активен
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="MobileClient"
      >
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="active"
        >
          Активен
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="MobileClient"
      >
        <td>
          Петров
        </td>
        <td>
          Пётр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="active"
        >
          Активен
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="MobileClient"
      >
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          220
        </td>
        <td
          className="active"
        >
          Активен
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
  <div
    className="Card"
  />
</div>
`;

exports[`работа MobileCompany 2`] = `
<div
  className="MobileCompany"
>
  <input
    onClick={[Function]}
    type="button"
    value="Velcom"
  />
  <input
    onClick={[Function]}
    type="button"
    value="МТС"
  />
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <input
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="MobileClient"
      >
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="active"
        >
          Активен
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="MobileClient"
      >
        <td>
          Петров
        </td>
        <td>
          Пётр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="active"
        >
          Активен
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="MobileClient"
      >
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          220
        </td>
        <td
          className="active"
        >
          Активен
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
  <div
    className="Card"
  />
</div>
`;

exports[`работа MobileCompany 3`] = `
<div
  className="MobileCompany"
>
  <input
    onClick={[Function]}
    type="button"
    value="Velcom"
  />
  <input
    onClick={[Function]}
    type="button"
    value="МТС"
  />
  <div
    className="MobileCompanyName"
  >
    Компания «
    МТС
    »
  </div>
  <input
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="MobileClient"
      >
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="active"
        >
          Активен
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="MobileClient"
      >
        <td>
          Петров
        </td>
        <td>
          Пётр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="active"
        >
          Активен
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="MobileClient"
      >
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          220
        </td>
        <td
          className="active"
        >
          Активен
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
  <div
    className="Card"
  />
</div>
`;

exports[`работа MobileCompany 4`] = `
<div
  className="MobileCompany"
>
  <input
    onClick={[Function]}
    type="button"
    value="Velcom"
  />
  <input
    onClick={[Function]}
    type="button"
    value="МТС"
  />
  <div
    className="MobileCompanyName"
  >
    Компания «
    МТС
    »
  </div>
  <input
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
  <div
    className="Card"
  />
</div>
`;
